File: P2_Regression——回归分析/multiple_linear.py
# Multiple linear Regression #

# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing the dataset
dataset = pd.read_csv('50_Startups.csv')
X = dataset.iloc[:, :-1].values
y = dataset.iloc[:, 4].values
# encode categorical data #
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
labelencoder_X = LabelEncoder()
X[:,3]=labelencoder_X.fit_transform(X[:,3])
onehotencoder = OneHotEncoder(categorical_features =[3])
X = onehotencoder.fit_transform(X).toarray()

# aviod dummy variable trap
X = X[:,1:]


# Splitting the dataset into the Training set and Test set
from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.20, random_state = 0)

# Fitting multiple linear regression to the Traning set
from sklearn.linear_model import LinearRegression
regressor = LinearRegression()
regressor.fit(X_train, y_train)

# Predicting the Test set results
y_pred = regressor.predict(X_test)

# Building the optimal using Backwards Elimination
import statsmodels.formula.api as sm
# add constance vector into X
# if not add astype(int),np.ones will give type error
X = np.append(arr = X,values = np.ones((50,1)).astype(int), axis =1)
# if we want ones col is the first column in X, we need to reverse the arr
X = np.append(arr = np.ones((50,1)).astype(int),values = X, axis =1)
X_opt = X[:,[0,1,2,3,4,5]]
regressor_OLS = sm.OLS(endog = y, exog = X_opt).fit()
regressor_OLS.summary()
# look at highest p-value, remove x2
X_opt = X[:,[0,1,3,4,5]]
regressor_OLS = sm.OLS(endog = y, exog = X_opt).fit()
regressor_OLS.summary()
# remove variable with highest p-value, x1
X_opt = X[:,[0,3,4,5]]
regressor_OLS = sm.OLS(endog = y, exog = X_opt).fit()
regressor_OLS.summary()
# remove x2
X_opt = X[:,[0,3,5]]
regressor_OLS = sm.OLS(endog = y, exog = X_opt).fit()
regressor_OLS.summary()

X_opt = X[:,[0,3]]
regressor_OLS = sm.OLS(endog = y, exog = X_opt).fit()
regressor_OLS.summary()



 

File: P2_Regression——回归分析/multiple_linear.R
# Multiple linear regression

# Importing the dataset
dataset = read.csv('50_Startups.csv')

# encode categorical data
dataset$State = factor(dataset$State,
                         levels = c('New York', 'California','Florida'),
                         labels = c(1,2,3))


# Splitting the dataset into the Training set and Test set
# install.packages('caTools')
#library(caTools)
set.seed(123)
split = sample.split(dataset$Profit, SplitRatio = 0.8)
training_set = subset(dataset, split == TRUE)
test_set = subset(dataset, split == FALSE)

# fitting multiple linear regression to the training set
#regressor = lm(formula = Profit ~ R.D.Spend + Administration + Marketing.Spend + State )
regressor = lm(formula = Profit ~ . ,
               data = training_set)

summary(regressor)

# predicting the test set results
y_pred = predict(regressor, newdata = test_set)

# building the optimal model using Backwards Elimination
# no need to add constant vector to model as it is done by R
regressor = lm(formula = Profit ~ R.D.Spend + Administration + Marketing.Spend + State,
               data = dataset)

summary(regressor)

# remove State
regressor = lm(formula = Profit ~ R.D.Spend + Administration + Marketing.Spend,
               data = dataset)

summary(regressor)

# remove administration
regressor = lm(formula = Profit ~ R.D.Spend + Marketing.Spend,
               data = dataset)

summary(regressor)

# remove marketing.spend but p-value of marketing.spend already decreased to 0.06
# so it may not remove marketing.spending
regressor = lm(formula = Profit ~ R.D.Spend,
               data = dataset)

summary(regressor)



